sd/endpointer.go

```diff
@@ -0,0 +1,75 @@
+package sd
+
+import (
+	"io"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/log"
+	iEndpoint "github.com/go-kit/kit/sd/internal/endpoint"
+)
+
+// Endpointer listens to a service discovery system and yields a set of
+// identical endpoints on demand. An error indicates a problem with connectivity
+// to the service discovery system, or within the system itself; an Endpointer
+// may yield no endpoints without error.
+type Endpointer interface {
+	Endpoints() ([]endpoint.Endpoint, error)
```

Comment:
I had typed a more thorough reply but in the end it boils down to: LGTM. So,

> `Event` struct . . .

Yep, LGTM.

> I am back to my suggestion of removing the `error` from the `Endpointer` signature

Also SGTM to remove the error from the Endpointer interface. It means you must choose what to do when the endpointCache receives a non-nil error Event from the Instancer. The most obvious choice is to leave the current behavior and immediately invalidate all of the Endpoints. If you want to do something smarter, like only invalidate them after a timeout, you'll need to parameterize that behavior in the endpointCache constructor. Without having made prototype implementations to see how each option feels, I don't have strong opinions which choice is better.